src/kvstore/RocksEngine.h

```diff
@@ -92,9 +117,26 @@ class RocksPrefixIter : public KVIterator {
     return folly::StringPiece(iter_->value().data(), iter_->value().size());
   }
 
+  rocksdb::Slice* iterateUpperBound() const {
+    return iterateUpperBound_.get();
+  }
+
+  void setIterator(rocksdb::Iterator* iter) {
+    iter_.reset(iter);
+  }
+
+  void setIterator(std::unique_ptr<rocksdb::Iterator> iter) {
+    iter_ = std::move(iter);
+  }
+
  protected:
-  std::unique_ptr<rocksdb::Iterator> iter_;
+  std::unique_ptr<rocksdb::Iterator> iter_{nullptr};
   rocksdb::Slice prefix_;
+  // temporary key to make sure data_ of iterateUpperBound_ valid during its lifetime
+  std::string iterateUpperBoundKey_;
```

Comment:
Since it's temporary, I prefer keeping its lifetime temp as well, that is, not setting as protected/private variable.

Reply:
iterateUpperBoundKey_ is not *really* temporary. This string should be valid during the whole lifetime of this iterator. Otherwise, if the string got destroyed before iterator, the iterate_upper_bound of rocksdb::ReadOptions will become invalid, which will then result in crash.

Reply:
then the field `iterateUpperBound_` is no needed, you could construct `rocksdb::Slice` from upperBoundKey_ every time when needed. Since rocksdb::Slice is an string view, and really lightweight.

Reply:
iterateUpperBound_ is only constructed the first time when set ReadOptions, not every time..

Reply:
updated. We passed a pointer to ReadOptions and we should make sure the pointer valid before the iterator is destructed, so the Slice also should not be a local temporary variable.